src/sage/combinat/SJT.py

```diff
@@ -0,0 +1,255 @@
+        sage: s = SJT([1, 2, 3, 4]); s
+        [1, 2, 3, 4]
+        sage: s = SJT([1]); s
+        [1]
+        sage: s = s.next(); s
+        False
+        sage: s = SJT([]); s
+        []
+        sage: s = s.next(); s
+        False
+    """
+    def __init__(self, l, directions=None) -> None:
+        r"""
+        Transpose two elements at positions ``a`` and ``b`` in ``perm`` and
+        their corresponding directions as well following the
+        Steinhaus-Johnson-Trotter algorithm.
+
+        Each permutation is obtained by transposing two adjacent elements from
+        the previous permutation.
+
+        INPUT:
+
+        - ``l`` -- list; a list of ordered ``int``.
+
+        - ``directions`` -- list (default: ``None`` ); a list of directions for
+          each element in the permuted list. Used when constructing permutations
+          from a pre-defined internal state.
+
+        EXAMPLES::
+
+            sage: from sage.combinat.SJT import SJT
+            sage: s = SJT([1, 2, 3, 4]); s
+            [1, 2, 3, 4]
+            sage: s = s.next(); s
+            [1, 2, 4, 3]
+            sage: p = Permutation(s._get_perm(), algorithm='sjt', sjt=s)
+            sage: p
+            [1, 2, 4, 3]
+            sage: p.next()
+            [1, 4, 2, 3]
+
+        TESTS::
+
+            sage: from sage.combinat.SJT import SJT
+            sage: s = SJT([1, 3, 2, 4])
+            Traceback (most recent call last):
+            ...
+            ValueError: no internal state directions were given for non-identity
+            starting permutation for Steinhaus-Johnson-Trotter algorithm
+            sage: s = SJT([]); s
+            []
+            sage: s = s.next(); s
+            False
+        """
+        # The permuted list.
+        self._list = l
+
+        # The length of the permuted list. Return early on empty list.
+        self._n = len(l)
+        if self._n == 0:
+            return
+
+        if directions is None:
+            if not all(l[i] <= l[i+1] for i in range(self._n - 1)):
+                raise ValueError("no internal state directions were given for "
+                "non-identity starting permutation for "
+                "Steinhaus-Johnson-Trotter algorithm")
+            self._directions = [-1] * self._n
+
+            # The first element has null direction.
+            self._directions[0] = 0
+        else:
+            self._directions = directions
+
+    def __idx_largest_element_non_zero_direction(self, perm, directions):
+        r"""
+        Find the largest element in ``perm`` with a non null direction.
+        """
+        largest = 0
+        index = None
+        for i in range(self._n):
+            if directions[i] != 0:
+                e = perm[i]
+                if e > largest:
+                    index = i
+                    largest = e
+
+        return index
+
+    def _get_perm(self):
+        r"""
+        Return the current permutation of ``self``.
```

Comment:
I don't understand this method. If `_get_perm` is the same as `_list`, plus both are hidden variables (`_...`), why do we need it? Just use `._list`.

Reply:
No that's my bad, initially it was a hidden variable with `__` prefix but your suggestions to align with `permutation.py` changed it but it stayed private in my mind.

Commit 1052244ab774d16e727e38b999cbc93068319b0a should fix that